### PR TITLE
Add generic constraint function `annotate_constraint_groupings()`

### PR DESCRIPTION
**Pull request by averywpx**

### CODE REVIEW

File: gnomad/utils/constraint.py [hunk 1]
```diff
@@ -360,3 +362,76 @@ def collapse_strand(
         if isinstance(t, hl.Table)
         else t.annotate_rows(**collapse_expr)
     )
+
+
+def annotate_constraint_groupings(
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:367]: I'm wondering if this should be split into two functions.
One that is just the explode, so sort of like this:
```
def explode_by_vep_annotation(
    t: Union[hl.Table, hl.MatrixTable],
    vep_annotation: str,
    vep_root: str = "vep",
) -> Tuple[Union[hl.Table, hl.MatrixTable], Tuple[str]]:
    # Annotate 'worst_csq_by_gene' to t if t is used to build the "worst_csq" model.
    if vep_annotation == "worst_csq_by_gene":
        t = process_consequences(t)

    if vep_annotation not in t[vep_root].keys():
        raise ValueError(
            f"{vep_annotation} is not a row field of the VEP annotation in Table"
        )
    # Create top-level annotation for `vep_annotation` by pulling out from `vep_root`
    # annotation and explode the `vep_annotation`.
    if isinstance(t, hl.Table):
        t = t.transmute(**{vep_annotation: t[vep_root][vep_annotation]})
        t = t.explode(t[vep_annotation])
    else:
        t = t.transmute_rows(**{vep_annotation: t[vep_root][vep_annotation]})
        t = t.explode_rows(t[vep_annotation])

    return t
```

And one that is basically just to get the groupings so something like this:
```
def get_constraint_grouping_expr(
    vep_annotation_expr,
    coverage_expr = None,
    include_transcript_group: bool = True,
    include_canonical_group: bool = True,
):
    lof_expr = vep_annotation_expr.lof
    polyphen_prediction_expr = vep_annotation_expr.polyphen_prediction

    # Create constraint annotations to be used for groupings.
    groupings = {
        "annotation": vep_annotation_expr.most_severe_consequence,
        "modifier": hl.coalesce(lof_expr, polyphen_prediction_expr, "None"),
        "gene": vep_annotation_expr.gene_symbol,
    }
    if coverage_expr is not None:
        groupings["coverage"] = coverage_expr

    # Add 'transcript' and 'canonical' annotation if requested
    if include_transcript_group:
        groupings["transcript"] = vep_annotation_expr.transcript_id
    if include_canonical_group:
        groupings["canonical"] = hl.or_else(vep_annotation_expr.canonical == 1, False)

    return groupings
```

Then it adds some lines for their use, but I think it's better to split it like this to make it more generalizable.

Then add a function like this to the gnomad-constraint repo:
```
def annotate_constraint_groupings(ht, vep_annotation="transcript_consequences"):
    ht = explode_by_vep_annotation(ht, vep_annotation)
    if vep_annotation != "transcript_consequences":
        include_transcript_group = False
        include_canonical_group = False
    groupings = get_constraint_grouping_expr(
        ht[vep_annotation],
        coverage_expr=ht.exome_coverage,
        include_transcript_group=include_transcript_group,
        include_canonical_group=include_canonical_group,
    )

    return ht.annotate(**groupings), tuple(groupings.keys())
```

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
Some small changes, and a thought on possible restructuring to make it a bit more generalized

File: gnomad/utils/constraint.py [hunk 2]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
     return pop_lengths
+
+
+def explode_by_vep_annotation(
+    t: Union[hl.Table, hl.MatrixTable],
+    vep_annotation: str,
+    vep_root: str = "vep",
+) -> Union[hl.Table, hl.MatrixTable]:
+    """
+    Annotate specified VEP annotation if it's not in the input Table or MatrixTable and explode the VEP annotation.
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:715]:
```suggestion
    Explode the specified VEP annotation on the input Table/MatrixTable.
```

File: gnomad/utils/constraint.py [hunk 3]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+
+    :param t: Input Table or MatrixTable.
+    :param vep_annotation: Name of annotation in VEP annotation that will be annotated,
+        if necessary, and explode.
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:719]:
```suggestion
    :param vep_annotation: Name of annotation in `vep_root` to explode.
```

File: gnomad/utils/constraint.py [hunk 4]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+    :param vep_root: Name used for VEP annotation. Default is 'vep'.
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:720]:
```suggestion
    :param vep_root: Name used for root VEP annotation. Default is 'vep'.
```

File: gnomad/utils/constraint.py [hunk 5]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+    """
+    # Annotate 'worst_csq_by_gene' to t if it's specified for `vep_annotation`.
+    if vep_annotation == "worst_csq_by_gene":
+        t = process_consequences(t)
+
+    if vep_annotation not in t[vep_root].keys():
+        raise ValueError(
+            f"{vep_annotation} is not a row field of the VEP annotation in"
+            " Table/MatrixTable!"
+        )
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:731]:
```suggestion
            f"{vep_annotation} is not a row field of the {vep_root} annotation in"
            " Table/MatrixTable!"
        )
```

File: gnomad/utils/constraint.py [hunk 6]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+    include_transcript_group: bool = True,
+    include_canonical_group: bool = True,
+) -> Dict[str, Union[hl.StringExpression, hl.Int32Expression, hl.BooleanExpression]]:
+    """
+    Collect annotations used for constraint groupings.
+
+    Function collects the following annotations:
+        - annotation - 'most_severe_consequence' annotation in `vep_annotation`
+        - modifier - classic lof annotation, LOFTEE annotation, or PolyPhen annotation
+          in `vep_annotation`
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:755]: Or "None" if neither is defined. Also specify that the first two are pulled from the 'lof' annotation on `vep_annotation_expr` and the last is from the 'polyphen_prediction' annotation on `vep_annotation_expr`

`vep_annotation` no longer exists

File: gnomad/utils/constraint.py [hunk 7]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+    vep_annotation_expr: hl.StructExpression,
+    coverage_expr: hl.Int32Expression = None,
+    include_transcript_group: bool = True,
+    include_canonical_group: bool = True,
+) -> Dict[str, Union[hl.StringExpression, hl.Int32Expression, hl.BooleanExpression]]:
+    """
+    Collect annotations used for constraint groupings.
+
+    Function collects the following annotations:
+        - annotation - 'most_severe_consequence' annotation in `vep_annotation`
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:753]: `vep_annotation` no longer exists

File: gnomad/utils/constraint.py [hunk 8]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+        - modifier - classic lof annotation, LOFTEE annotation, or PolyPhen annotation
+          in `vep_annotation`
+        - gene - 'gene_symbol' annotation inside `vep_annotation`
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:756]: `vep_annotation` no longer exists

File: gnomad/utils/constraint.py [hunk 9]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+        - coverage - exome coverage
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:757]:
```suggestion
        - coverage - exome coverage if `coverage_expr` is specified
```

File: gnomad/utils/constraint.py [hunk 10]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+        - transcript (added when `include_transcript_group` is True)
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:758]: 'transcript_id' annotation on `vep_annotation_expr`

File: gnomad/utils/constraint.py [hunk 11]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+        - canonical (added when `include_canonical_group` is True)
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:759]: based on 'canonical' annotation on `vep_annotation_expr`

File: gnomad/utils/constraint.py [hunk 12]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+    else:
+        t = t.transmute_rows(**{vep_annotation: t[vep_root][vep_annotation]})
+        t = t.explode_rows(t[vep_annotation])
+
+    return t
+
+
+def get_constraint_grouping_expr(
+    vep_annotation_expr: hl.StructExpression,
+    coverage_expr: hl.Int32Expression = None,
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:745]:
```suggestion
    coverage_expr: Optional[hl.Int32Expression] = None,
```

File: gnomad/utils/constraint.py [hunk 13]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+        `vep_annotation_expr`:
+        - lof
+        - polyphen_prediction
+        - most_severe_consequence
+        - gene_symbol
+        - transcript_id (if `include_transcript_group` is True)
+        - canonical (if `include_canonical_group` is True)
+
+    :param vep_annotation_expr: StructExpression of VEP annotation.
+    :param coverage_expr: Int32Expression of exome coverage. Default is None.
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:772]:
```suggestion
    :param coverage_expr: Optional Int32Expression of exome coverage. Default is None.
```

File: gnomad/utils/constraint.py [hunk 14]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+    :param include_transcript_group: Wether to include the transcript annotation in the
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:773]:
```suggestion
    :param include_transcript_group: Whether to include the transcript annotation in the
```

File: gnomad/utils/constraint.py [hunk 15]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+        groupings. Default is True.
+    :param include_canonical_group: Wether to include canonical annotation in the
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:775]:
```suggestion
    :param include_canonical_group: Whether to include canonical annotation in the
```

File: gnomad/utils/constraint.py [hunk 16]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+        groupings["canonical"] = hl.or_else(vep_annotation_expr.canonical == 1, False)
+
+    return groupings
+
+
+def annotate_exploded_vep_for_constraint_groupings(
+    ht: hl.Table, vep_annotation: str = "transcript_consequences"
+) -> Tuple[Union[hl.Table, hl.MatrixTable], Tuple[str]]:
+    """
+    Annotate constraint annotations used for groupings.
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:805]:
```suggestion
    Annotate Table with annotations used for constraint groupings.
```

File: gnomad/utils/constraint.py [hunk 17]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+
+    Function explodes the specified VEP annotation and annotates the following
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:807]:
```suggestion
    Function explodes the specified VEP annotation (`vep_annotation`) and adds the following
```

File: gnomad/utils/constraint.py [hunk 18]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+        - canonical (added when `vep_annotation` is specified as
+          "transcript_consequences")
+
+    .. note::
+        This function expects that the following fields are present in `ht`:
+        - vep
+        - exome_coverage
+
+    :param t: Input Table or MatrixTable.
+    :param vep_annotation: Name of annotation in VEP annotation (one of
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:825]:
```suggestion
    :param vep_annotation: Name of annotation in 'vep' annotation (one of
```

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
Looking good, just some minor changes

File: gnomad/utils/constraint.py [hunk 19]
```diff
@@ -700,3 +704,207 @@ def get_all_pop_lengths(
...
+    pop: Optional[str] = None,
+) -> Dict[str, Union[hl.Float64Expression, hl.Int64Expression]]:
+    """
+    Apply plateau models for all sites and for a population (if specified) to compute predicted proportion observed ratio and expected variant counts.
+
+    .. note:
+        Function expects following annotations to be present in `ht`:
+        - cpg
+        - observed_variants
+        - possible_variants
```
[the remark this answers is not in the document]
Comment (averywpx) [reply gnomad/utils/constraint.py:863]: Yes I agree. I used to think that `cpg` is not obvious from function name, but it won't be necessary after I switched to `cpg_expr`.

File: gnomad/utils/constraint.py [hunk 20]
```diff
@@ -700,3 +707,177 @@ def get_all_pop_lengths(
...
+    """
+    Annotate Table with annotations used for constraint groupings.
+
+    Function explodes the specified VEP annotation (`vep_annotation`) and adds the following
+    annotations:
+    - annotation -'most_severe_consequence' annotation in `vep_annotation`
+    - modifier - classic lof annotation from 'lof' annotation in
+        `vep_annotation_expr`, LOFTEE annotation from 'lof' annotation in
+        `vep_annotation_expr`, PolyPhen annotation from 'polyphen_prediction' in
+        `vep_annotation_expr`, or "None" if neither is defined
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:786]:
```suggestion
        `vep_annotation`, LOFTEE annotation from 'lof' annotation in
        `vep_annotation`, PolyPhen annotation from 'polyphen_prediction' in
        `vep_annotation`, or "None" if neither is defined
```

File: gnomad/utils/constraint.py [hunk 21]
```diff
@@ -700,3 +707,177 @@ def get_all_pop_lengths(
...
+    - gene - 'gene_symbol' annotation inside `vep_annotation`
+    - coverage - exome coverage in `t`
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:788]:
```suggestion
    - coverage - exome coverage in `ht`
```

File: gnomad/utils/constraint.py [hunk 22]
```diff
@@ -700,3 +707,177 @@ def get_all_pop_lengths(
...
+    - transcript - id from 'transcript_id' in `vep_annotation_expr` (added when
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:789]:
```suggestion
    - transcript - id from 'transcript_id' in `vep_annotation` (added when
```

File: gnomad/utils/constraint.py [hunk 23]
```diff
@@ -700,3 +707,177 @@ def get_all_pop_lengths(
...
+        `include_transcript_group` is True)
+    - canonical from `vep_annotation_expr` (added when `include_canonical_group` is
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:791]:
```suggestion
    - canonical from `vep_annotation` (added when `include_canonical_group` is
```

File: gnomad/utils/constraint.py [hunk 24]
```diff
@@ -700,3 +707,177 @@ def get_all_pop_lengths(
...
+    mu_expr: hl.Float64Expression,
+    cov_corr_expr: hl.Float64Expression,
+    cpg_expr: hl.BooleanExpression,
+    pop: Optional[str] = None,
+) -> Dict[str, Union[hl.Float64Expression, hl.Int64Expression]]:
+    """
+    Apply plateau models for all sites and for a population (if specified) to compute predicted proportion observed ratio and expected variant counts.
+
+    :param ht: Input Table.
+    :param plateau_models: Linear models (output of `build_models()`, with the values
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:841]:
```suggestion
    :param plateau_models_expr: Linear models (output of `build_models()`, with the values
```

File: gnomad/utils/constraint.py [hunk 25]
```diff
@@ -700,3 +707,177 @@ def get_all_pop_lengths(
...
+
+    # Generate sum aggregators for 'predicted_proportion_observed' and
+    # 'expected_variants', for specified population.
+    agg_expr = {
+        f"predicted_proportion_observed{pop}": agg_func(ppo_expr),
+        f"expected_variants{pop}": agg_func(ppo_expr * cov_corr_expr),
+    }
+
+    # Generate sum aggregators for 'observed_variants' and 'possible_variants' on
+    # the entire dataset if pop is None, and for `downsampling_counts` for
```
Comment (jkgoodrich) [inline gnomad/utils/constraint.py:879]:
```suggestion
    # the entire dataset if pop is None, and for 'downsampling_counts' for
```

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
A couple small docstring changes. You also need to run Black and fix the failing docstring test.

File: gnomad/utils/constraint.py (hunk 20 above)
Comment (averywpx) [reply gnomad/utils/constraint.py]: I remember there was a comment from last round of review:

> Or "None" if neither is defined. Also specify that the first two are pulled from the 'lof' annotation on vep_annotation_expr and the last is from the 'polyphen_prediction' annotation on vep_annotation_expr
> 
> vep_annotation no longer exists

I made the change and just want to double check it is desired!

Comment (jkgoodrich) [reply gnomad/utils/constraint.py]: Yup, that change looks good to me. Thanks!

**Pull request merged**